Address removals of doctrine-dbal

MarcBrillault opened the pull request:
- `Doctrine\DBAL\Schema\Visitor\RemoveNamespacedAssets`
- `Doctrine\DBAL\Schema\Schema::visit`
- `Doctrine\DBAL\Tools\Console\Command\ReservedWordsCommand`

MarcBrillault committed:
Address removals of doctrine-dbal

- Doctrine\DBAL\Schema\Visitor\RemoveNamespacedAssets
- Doctrine\DBAL\Schema\Schema::visit
- Doctrine\DBAL\Tools\Console\Command\ReservedWordsCommand

greg0ire committed:
Run foreign key tests on all platforms

SQLite now supports foreign keys.

morozov commented:
> 🤔 looks like there are remaining failures for MySQL 😞

That's the bug that the `RemoveNamespacedAssets` visitor is no longer causing. The new expected assertions should be like this:
```php
self::assertEquals('CREATE TABLE gh6823_user (id VARCHAR(255) NOT NULL, group_id VARCHAR(255) CHARACTER SET ascii DEFAULT NULL COLLATE `ascii_general_ci`, status_id VARCHAR(255) CHARACTER SET latin1 DEFAULT NULL COLLATE `latin1_bin`, INDEX IDX_70DD1774FE54D947 (group_id), INDEX IDX_70DD17746BF700BD (status_id), PRIMARY KEY(id)) DEFAULT CHARACTER SET utf8mb4 COLLATE `utf8mb4_bin` ENGINE = InnoDB', $this->getLastLoggedQuery(7)['sql']);
self::assertEquals('CREATE TABLE gh6823_user_tags (user_id VARCHAR(255) CHARACTER SET utf8mb4 NOT NULL COLLATE `utf8mb4_bin`, tag_id VARCHAR(255) CHARACTER SET latin1 NOT NULL COLLATE `latin1_bin`, INDEX IDX_596B1281A76ED395 (user_id), INDEX IDX_596B1281BAD26311 (tag_id), PRIMARY KEY(user_id, tag_id)) DEFAULT CHARACTER SET ascii COLLATE `ascii_general_ci` ENGINE = InnoDB', $this->getLastLoggedQuery(6)['sql']);
self::assertEquals('CREATE TABLE gh6823_group (id VARCHAR(255) NOT NULL, PRIMARY KEY(id)) DEFAULT CHARACTER SET ascii COLLATE `ascii_general_ci` ENGINE = InnoDB', $this->getLastLoggedQuery(5)['sql']);
self::assertEquals('CREATE TABLE gh6823_status (id VARCHAR(255) CHARACTER SET latin1 NOT NULL COLLATE `latin1_bin`, PRIMARY KEY(id)) DEFAULT CHARACTER SET koi8r COLLATE `koi8r_bin` ENGINE = InnoDB', $this->getLastLoggedQuery(4)['sql']);
self::assertEquals('ALTER TABLE gh6823_user ADD CONSTRAINT FK_70DD1774FE54D947 FOREIGN KEY (group_id) REFERENCES gh6823_group (id)', $this->getLastLoggedQuery(3)['sql']);
self::assertEquals('ALTER TABLE gh6823_user ADD CONSTRAINT FK_70DD17746BF700BD FOREIGN KEY (status_id) REFERENCES gh6823_status (id)', $this->getLastLoggedQuery(2)['sql']);
self::assertEquals('ALTER TABLE gh6823_user_tags ADD CONSTRAINT FK_596B1281A76ED395 FOREIGN KEY (user_id) REFERENCES gh6823_user (id)', $this->getLastLoggedQuery(1)['sql']);
self::assertEquals('ALTER TABLE gh6823_user_tags ADD CONSTRAINT FK_596B1281BAD26311 FOREIGN KEY (tag_id) REFERENCES gh6823_tag (id)', $this->getLastLoggedQuery()['sql']);
```
But if you want to test both with and without the visitor, this test should be completely reworked since it's too strict. Furthermore, with the visitor in place, it considers the bug in the visitor as the expected behavior which is wrong.

greg0ire committed:
Use a better check for the DBAL version

It seems AbstractPlatform::getGuidExpression() cannot do the trick, as
it is missing from both branches.

greg0ire committed:
Run test only with DBAL 3

With DBAL 4, there is no longer a filtering of assets.

morozov commented on tests/Doctrine/Tests/ORM/Functional/Ticket/GH6823Test.php:
```diff
@@ -27,7 +27,7 @@ public function testCharsetCollationWhenCreatingForeignRelations(): void
 
-        if (method_exists(AbstractPlatform::class, 'getGuidExpression')) {
+        if (! class_exists(Visitor::class)) {
             self::markTestSkipped('Test valid for doctrine/dbal:3.x only.');
```
This is a bit inaccurate: the scenario being tested doesn't depend on the Visitor and is valid for DBAL 4 as well. I'd rather have two different sets of assertions depending on whether the visitor exists or not.

Once the support for DBAL 3 is dropped, only one branch will be left. Otherwise, most likely this test will be mechanically removed as no longer relevant.

greg0ire replied:
Sure, I can work on that.

greg0ire replied:
Note that I'm only using `Visitor` in order to detect the version, I don't really care about that class.

greg0ire replied:
I'm rate limited by GH's API, can we maybe merge as is? I'll treat this separately on my own fork. As is, the test is fixed to behave as advertised in "Test valid for doctrine/dbal:3.x only."

morozov replied:
> Note that I'm only using `Visitor` in order to detect the version, I don't really care about that class.

You should be checking for `RemoveNamespacedAssets` instead since the result of the test does depend on whether it exists or not.

greg0ire replied:
Done.

greg0ire commented:
Thanks @MarcBrillault !
